### ISSUE TITLE
Directly passing Rc::clone(...) to a function expecting Rc<dyn Trait> produces unexpected type error

### ISSUE THREAD

**Issue by periodicaidan** [rust-lang/rust#75240]
### Summary

Given some struct `S` that implements a trait `T`, attempting to pass `Rc::clone(&s)` (where `s` has type `Rc<S>`) as an argument that should have type `Rc<dyn T>` produces the compiler error, "expected trait object `dyn T`; found struct `S`". However, passing simply `s`, or storing the cloned `Rc` in a variable and passing that, produces no such error.

### Example

Below is a trivial example that illustrates how to achieve this particular compiler error. ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=87d836d6563b9b86501a429ae312aec9))

```rust
use std::rc::Rc;

// Readable trait
trait Readable {
    fn read(&self) -> String;
}

// Poem implements the Readable trait
struct Poem {
    pub stanza_number: usize,
    pub stanzas: Vec<String>
}

impl Poem {
    pub fn new(stanzas: Vec<&str>) -> Self {
        Self {
            stanza_number: 0,
            stanzas: stanzas.iter().map(|s| s.to_string()).collect()
        }
    }
}

impl Readable for Poem {
    fn read(&self) -> String {
        self.stanzas[self.stanza_number].clone()
    }
}

// A Library holds a collection of structs that implement Readable
struct Library {
    pub collection: Vec<Rc<dyn Readable>>
}

fn main() {
    let erlkoenig = Rc::new(Poem::new(vec![
"Who rides, so late, through night and wind?\n
It is the father with his child.\n
He has the boy well in his arm\n
He holds him safely, he keeps him warm.",

// ...
 ]));
    
    let goethes_works = Library {
        collection: vec![
            Rc::clone(&erlkoenig)
        ]
    };
}
```

I would expect this code to compile normally because `erlkoenig` is an instance of `Poem`, which implements `Readable`. However it fails to compile with the following error:

```
error[E0308]: mismatched types
  --> src/main.rs:77:23
   |
77 |             Rc::clone(&erlkoenig)
   |                       ^^^^^^^^^^ expected trait object `dyn Readable`, found struct `Poem`
   |
   = note: expected reference `&std::rc::Rc<dyn Readable>`
              found reference `&std::rc::Rc<Poem>`
```

On the other hand, the code does compile if you remove the `Rc::clone(...)`:

```rust
fn main() {
    let erlkoenig = Rc::new(Poem::new(vec![/* ... */]));
    
    let goethes_works = Library {
        collection: vec![
            erlkoenig
        ]
    };
}
```

Or if the clone is stored in a variable:

```rust
fn main() {
    let erlkoenig = Rc::new(Poem::new(vec![/* ... */]));

    let erl_clone_ig = Rc::clone(&erlkoenig);
    
    let goethes_works = Library {
        collection: vec![
            erl_clone_ig
        ]
    };
}
```

### Meta

I first encountered this issue trying to compile on nightly (1.47.0-nightly: 6c8927b0c 2020-07-26).

I have tried this code in the Playground on stable (1.45.1), beta (1.46.0-beta.2: 2020-07-23 6f959902b3103c49ca98), and nightly (1.47.0-nightly: 2020-08-02 81e754c359c471f91263). It fails to compile in any of them and emits the same error.

**Comment by nbdd0121** [comment]
This is expected behaviour.

* Firstly you need to know that `Rc<Poem>` can converted to `Rc<dyn Readable>` by unsize coercion. However, `&Rc<Poem>` cannot be converted to `&Rc<dyn Readable>` via unsize coercion (this is double pointer already).
* Secondly, when you write `Rc::clone`, you actually ask for `<Rc<_> as Clone>::clone`, which involves type inference. Because coercion is not performed whenever there is a type inference, `_` here can only be `dyn Readable`, but not `Poem`, so this means that you are writing `<Rc<dyn Readable> as Clone>::clone(&xxx)`, which requires `&xxx` to be `&dyn Readable`, and by the first bullet point, this is not part of unsize coercion.

If you write `Rc::<Poem>::clone(&xxx)`, this would be legal because the return type is not inferred, so an unsize coercion takes place. If you store the clone in a separate variable this is similar, because that variable will have type `Rc<Poem>` so an unsize coercion can happen when you give it to collection. The easiest way would be just to write `xxx.clone()` directly because that also gives you a `Rc<Poem> and permits unsize coercion.

You can read more about coercion on nomicon: https://doc.rust-lang.org/nomicon/coercions.html.

**Comment by periodicaidan** [comment]
Ah, thank you for the explanation and the resource!
